File: generated-code/codecraft/typescript/src/model/attack-properties.ts
import { StreamWrapper } from "../stream-wrapper";

/**
 * Entity's attack properties
 */
export class AttackProperties {
    /**
     * Maximum attack range
     */
    attackRange: number
    /**
     * Damage dealt in one tick
     */
    damage: number
    /**
     * If true, dealing damage will collect resource from target
     */
    collectResource: boolean

    constructor(attackRange: number, damage: number, collectResource: boolean) {
        this.attackRange = attackRange;
        this.damage = damage;
        this.collectResource = collectResource;
    }

    /**
     * Read AttackProperties from input stream
     */
    static async readFrom(stream: StreamWrapper): Promise<AttackProperties> {
        let attackRange;
        attackRange = await stream.readInt();
        let damage;
        damage = await stream.readInt();
        let collectResource;
        collectResource = await stream.readBool();
        return new AttackProperties(attackRange, damage, collectResource)
    }

    /**
     * Write AttackProperties to output stream
     */
    async writeTo(stream: StreamWrapper) {
        let attackRange = this.attackRange;
        await stream.writeInt(attackRange);
        let damage = this.damage;
        await stream.writeInt(damage);
        let collectResource = this.collectResource;
        await stream.writeBool(collectResource);
    }
}

File: generated-code/codecraft/ruby/model/attack_properties.rb
class AttackProperties
    attr_accessor :attack_range
    attr_accessor :damage
    attr_accessor :collect_resource

    def initialize(attack_range, damage, collect_resource)
        @attack_range = attack_range
        @damage = damage
        @collect_resource = collect_resource
    end

    def self.read_from(stream)
        attack_range = stream.read_int()
        damage = stream.read_int()
        collect_resource = stream.read_bool()
        AttackProperties.new(attack_range, damage, collect_resource)
    end

    def write_to(stream)
        stream.write_int(@attack_range)
        stream.write_int(@damage)
        stream.write_bool(@collect_resource)
    end

    def to_s
        string_result = "AttackProperties { "
        string_result += "attack_range: "
        string_result += @attack_range.to_s
        string_result += ", "
        string_result += "damage: "
        string_result += @damage.to_s
        string_result += ", "
        string_result += "collect_resource: "
        string_result += @collect_resource.to_s
        string_result += " }"
        string_result
    end

    def to_str
        to_s
    end
end